graph/structure/impl/RuleStructureImpl.java

```diff
@@ -107,6 +182,12 @@ public Buffered(SchemaGraph graph, StructureIID.Rule iid, String label, Conjunct
             super(graph, iid, label, when, then);
             this.isCommitted = new AtomicBoolean(false);
             setModified();
+            indexLabels();
+        }
+
+        private void indexLabels() {
```

Comment:
a buffered rule structure will index the type labels that are mentioned in the rule explicitly to prevent types from being undefined while the label is in use in a rule